types/convert.go

```diff
@@ -173,6 +173,48 @@ func ConvertFloatToUint(sc *stmtctx.StatementContext, fval float64, upperBound u
 	return uint64(val), nil
 }
 
+// ConvertDecimalToUint converts a decimal to an uint by string instead of float to avoid float overflow (#10181).
+func ConvertDecimalToUint(sc *stmtctx.StatementContext, d *MyDecimal, upperBound uint64, tp byte) (uint64, error) {
+	// since MyDecimal.ToString() doesn't use scientific notation now, it's safe to ignore it.
+	// TODO: handle scientific notation
+	str := string(d.ToString())
+	var intStr, fracStr string
+	p := strings.Index(str, ".")
+	if p == -1 {
+		intStr = str
+	} else {
+		intStr = str[:p]
+		fracStr = str[p+1:]
+	}
+	for len(intStr) > 0 && intStr[0] == '0' {
+		intStr = intStr[1:]
+	}
+	if sc.ShouldClipToZero() && intStr[0] == '-' {
```

Comment:
I assume that `len(intStr) > 0` here because `MyDecimal` returns a valid decimal string without scientific notation.
But I will let it support scientific notation for safety.